src/Diagnostics.CompilerHost/Middleware/CompilerRequestMiddleware.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using Diagnostics.Logger;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+
+namespace Diagnostics.CompilerHost.Middleware
+{
+    public class CompilerRequestMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly string _apiLoggerKey = "CompilerHost_ApiLogger";
+
+        public CompilerRequestMiddleware(RequestDelegate next)
+        {
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            BeginRequest_Handle(httpContext);
+
+            try
+            {
+                await _next(httpContext);
+            }
+            catch (Exception ex)
+            {
+                if (!httpContext.Response.HasStarted)
+                {
+                    httpContext.Response.Clear();
+                    httpContext.Response.StatusCode = (int)HttpStatusCode.InternalServerError;
+                }
+
+                LogException(httpContext, ex);
+            }
+            finally
+            {
+                EndRequest_Handle(httpContext);
+            }
+
+            return;
+        }
+
+        private void BeginRequest_Handle(HttpContext httpContext)
+        {
+            var logger = new ApiMetricsLogger(httpContext);
+            httpContext.Items.Add(_apiLoggerKey, logger);
+        }
+
+        private void EndRequest_Handle(HttpContext httpContext)
+        {
+            ApiMetricsLogger logger = (ApiMetricsLogger)httpContext.Items[_apiLoggerKey];
+            if (logger == null)
+            {
+                logger = new ApiMetricsLogger(httpContext);
+            }
+
+            logger.LogCompilerHostAPIMetrics(httpContext);
+        }
+
+        private void LogException(HttpContext context, Exception ex)
+        {
+            try
+            {
+                ApiMetricsLogger logger = (ApiMetricsLogger)context.Items[_apiLoggerKey];
+                if (logger == null)
+                {
+                    logger = new ApiMetricsLogger(context);
+                }
+
+                logger.LogCompilerHostUnhandledException(context, ex);
+            }
+            catch (Exception logEx)
+            {
+                DiagnosticsETWProvider.Instance.LogCompilerHostUnhandledException(
+                    string.Empty,
```

Comment:
Same as with the DiagnosticRequestMiddleware, track requestId for correlation purposes

Reply:
sure.. This code path will get executed only when there is an exception in ETW logging. (I dont expect this to get hit lot of times. but ya, I can put reqquest id there.)

Reply:
Ya i was thinking the same about the ETW logging too. But I guess in those edge cases maybe this will help